kate/grid/src/grid.rs

```diff
@@ -0,0 +1,294 @@
+use alloc::vec::Vec;
+
+use crate::Dimensions;
+
+pub trait Grid<A> {
+	fn width(&self) -> usize;
+	fn height(&self) -> usize;
+	fn dims(&self) -> &Dimensions;
+	fn inner(&self) -> &Vec<A>;
+	// x indexes within a row, y indexes within a column
+	// 0 <= x < width, 0 <= y < height
+	fn get(&self, x: usize, y: usize) -> Option<&A> {
+		let i = Self::coord_to_ind(self.dims(), x, y);
+		self.get_ind(i)
+	}
+	fn get_ind(&self, i: usize) -> Option<&A>;
+	fn ind_to_coord(dims: &Dimensions, i: usize) -> (usize, usize);
+	fn coord_to_ind(dims: &Dimensions, x: usize, y: usize) -> usize;
+}
+
+pub struct RowMajor<A> {
+	dims: Dimensions,
+	inner: Vec<A>,
+}
+
+pub struct ColumnMajor<A> {
+	dims: Dimensions,
+	inner: Vec<A>,
+}
+
+impl<A> Grid<A> for RowMajor<A> {
+	fn width(&self) -> usize {
+		self.dims.width()
+	}
+
+	fn height(&self) -> usize {
+		self.dims.height()
+	}
+
+	fn dims(&self) -> &Dimensions {
+		&self.dims
+	}
+
+	fn get_ind(&self, i: usize) -> Option<&A> {
+		self.inner.get(i)
+	}
+
+	fn ind_to_coord(dims: &Dimensions, i: usize) -> (usize, usize) {
+		(i % dims.width_nz(), i / dims.width_nz())
+	}
+
+	fn coord_to_ind(dims: &Dimensions, x: usize, y: usize) -> usize {
+		x.saturating_add(y.saturating_mul(dims.width()))
+	}
```

Comment:
Yeah that's fair. Because these are usizes you have to really mess up or be on a 16-bit platform to hit these errors, so adding loads of options makes things a little annoying but I'll refactor and see if it doesn't make things too bad.